File: ios/Sources/App/AppModel.swift
import Foundation
import SwiftUI

/// 全局状态。界面只读它，写入一律经过这里，保证 store 与 UI 不会各说各话。
@MainActor
@Observable
final class AppModel {
    // 用 AppModel.today 而不是 Self.today：类的存储属性初始化式里不能引用
    // 协变的 Self，编译器直接拒。
    var project = FieldProject(
        id: "local",
        name: "未选择项目",
        archivePath: "现场影像 / \(AppModel.today)"
    )
    var items: [CaptureItem] = []
    var tally: TransferTally = .zero
    /// 桌面端连接。还没做配对，先恒定离线——**不要为了界面好看假装在线**，
    /// 那会让人以为照片已经在往电脑走。
    var link = DesktopLink(isOnline: false, lastSyncedAt: nil, deviceName: nil)

    /// 当前账号。nil = 未登录，界面走登录页。
    var account: AccountUser?
    var isSignedIn: Bool { account != nil }
    /// 启动时先按 Keychain 里有没有会话决定初屏，避免登录页闪一下再跳走。
    private(set) var didRestore = false

    static var today: String {
        let f = DateFormatter(); f.dateFormat = "yyyy-MM-dd"
        return f.string(from: Date())
    }

    // MARK: - 账号

    /// 选中的归档目标。nil = 还没选，界面走项目选择页。
    /// 存 UserDefaults 就够——项目条目不是凭据，泄露它没有意义。
    /// （旧版存的是 Int 型云端项目 id，键 selectedProjectId；目录镜像的 key 是
    /// 桌面机本地 id，两个命名空间不通，升级后旧选择一律作废、回到选择页。）
    var selectedProject: RelayProject? {
        didSet {
            if let p = selectedProject, let data = try? JSONEncoder().encode(p) {
                UserDefaults.standard.set(data, forKey: "selectedRelayProject")
            } else {
                UserDefaults.standard.removeObject(forKey: "selectedRelayProject")
            }
        }
    }

    func didLogin(_ result: LoginResult) async {
        account = result.user
        await refresh()
    }

    func selectProject(_ p: RelayProject) async {
        selectedProject = p
        project = FieldProject(id: p.id, name: p.name,
                               archivePath: "现场影像 / \(AppModel.today)")
        await UploadQueue.shared.configure(project: p) { [weak self] in
            await self?.refresh()
        }
        await UploadQueue.shared.kick()
    }

    /// 拍完立刻踢一脚队列。不等用户手动点上传——现场没人会记得点。
    func kickUpload() {
        Task { await UploadQueue.shared.kick() }
    }

    func retryFailedUploads() {
        Task { await UploadQueue.shared.retryFailed() }
    }

    /// 单条重试。队列页每行都有，不用为了一条失败去点「全部重试」。
    func retry(_ item: CaptureItem) {
        Task {
            try? await EvidenceStore.shared.updateState(item.id, to: .waiting, progress: 0)
            await refresh()
            await UploadQueue.shared.kick()
        }
    }

    /// 切项目：清掉选择回到选择页。不动已拍的影像，它们仍属于原项目的队列。
    func clearProjectSelection() {
        selectedProject = nil
    }

    /// 未投递件的中转区到期时刻（键 clientMediaId 小写）。队列页做到期提醒用。
    var cloudExpiry: [String: Date] = [:]

    /// 查投递回执：停在中转区的影像被桌面端取走后改成「已抵达」。
    func checkDelivered() {
        Task {
            cloudExpiry = await UploadQueue.shared.checkDelivered()
            await refresh()
        }
    }

    func signOut() {
        API.shared.logout()
        account = nil
        // **不清本地影像。** 退出登录不等于放弃已经拍到的东西——
        // 现场是不可复现的，登出就删是灾难性的默认。
    }

    func refresh() async {
        do {
            let all = try await EvidenceStore.shared.loadAll()
            let t = try await EvidenceStore.shared.tally()
            items = all
            tally = t
        } catch {
            // 读不出来不该清空界面——宁可显示上一次的状态，也不要让用户以为照片没了
        }
    }

    func bootstrap() async {
        // Keychain 里有会话就直接进主界面。这里只做「有没有」的判断，
        // 会话是否还有效由第一次真实请求的 401 来发现——启动时多打一次
        // 网络请求会让离线开 App 卡在转圈，而离线拍照恰恰是这个 App 的主场景。
        if SessionStore.current != nil, account == nil {
            account = AccountUser(id: 0, username: "", displayName: "", avatarUrl: "", role: "USER")
        }
        // 旧版的 Int 键作废（命名空间已换），顺手清掉
        UserDefaults.standard.removeObject(forKey: "selectedProjectId")
        if let data = UserDefaults.standard.data(forKey: "selectedRelayProject"),
           let saved = try? JSONDecoder().decode(RelayProject.self, from: data) {
            selectedProject = saved
            project = FieldProject(id: saved.id, name: saved.name,
                                   archivePath: "现场影像 / \(AppModel.today)")
            await UploadQueue.shared.configure(project: saved) { [weak self] in
                await self?.refresh()
            }
        }
        didRestore = true
        try? await EvidenceStore.shared.sweepOrphans()
        await refresh()
        // 上次没传完的，启动就接着传；停在中转区的顺手查一次回执
        if selectedProject != nil {
            await UploadQueue.shared.kick()
            cloudExpiry = await UploadQueue.shared.checkDelivered()
            await refresh()
        }
        // 上传自愈心跳（dev-board#241）：滞留的「传输中」与失败件不能等用户来点，
        // 现场拍完手机就揣兜里了——每分钟自动续传，失败按退避重试
        Task {
            while !Task.isCancelled {
                try? await Task.sleep(for: .seconds(60))
                await UploadQueue.shared.autoKick()
            }
        }
    }

    func store(data: Data, kind: MediaKind, at: Date, location: (lat: Double, lon: Double, accuracy: Double)?) async {
        do {
            _ = try await EvidenceStore.shared.save(
                data: data, kind: kind, capturedAt: at,
                location: location, device: Device.facts
            )
            await refresh()
            kickUpload()
            await saveToAlbumIfEnabled()
        } catch {
            lastError = "保存失败：\(error.localizedDescription)"
        }
    }

    /// 存进系统相册（开关打开时）。**与上传彼此独立**——
    /// 相册存不进去不该影响上传，反之亦然，两条链任何一条断了另一条都要继续走。
    private func saveToAlbumIfEnabled() async {
        guard Prefs.saveToAlbum else { return }
        // 录音不进相册：PhotoKit 没有音频资产，混进来只会在这个循环里反复空转。
        let pending = items.filter { !$0.savedToAlbum && $0.kind != .audio }
        for item in pending {
            do {
                try await AlbumSaver.save(url: item.localURL, kind: item.kind)
                try? await EvidenceStore.shared.markSavedToAlbum(item.id)
            } catch {
                // 不打断、不弹窗。相册失败最常见的原因是没给权限，
                // 设置页里已经把这件事说清楚了，这里再弹一次是噪音。
                albumWarning = error.localizedDescription
            }
        }
        await refresh()
    }

    var albumWarning: String?
    var lastError: String?
}

File: android/app/src/main/kotlin/com/aiworkdeck/mobile/AppModel.kt
package com.aiworkdeck.mobile

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import androidx.work.WorkManager
import com.aiworkdeck.mobile.model.CaptureItem
import com.aiworkdeck.mobile.model.MediaKind
import com.aiworkdeck.mobile.model.RelayProject
import com.aiworkdeck.mobile.model.TransferState
import com.aiworkdeck.mobile.model.TransferTally
import com.aiworkdeck.mobile.services.AccountUser
import com.aiworkdeck.mobile.services.Loc
import com.aiworkdeck.mobile.services.LoginResult
import com.aiworkdeck.mobile.services.MediaUsage
import com.aiworkdeck.mobile.services.RecordingState
import com.aiworkdeck.mobile.services.ServiceLocator
import com.aiworkdeck.mobile.services.Unauthorized
import com.aiworkdeck.mobile.services.UploadWorker
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import java.io.File
import java.time.Instant
import java.util.logging.Logger

/** 主界面之上盖着的整屏。None = 只有取景器。 */
enum class Overlay { None, Library, Queue, Settings }

/**
 * 全局状态。界面只读它，写入一律经过这里，保证磁盘上的库与屏幕上的画面不会各说各话。
 * 镜像 iOS `AppModel`（ios/Sources/App/AppModel.swift），差别只在 Flow 与 ViewModel 的形制。
 *
 * 依赖从 [ServiceLocator] 取而不是构造参数注入：UploadWorker 也走同一份实例，两边必须是
 * 同一个队列、同一份库，否则会出现「界面显示待传、Worker 说没活干」。
 */
class AppModel(app: Application) : AndroidViewModel(app) {
    private val logger = Logger.getLogger("AppModel")
    private val evidence = ServiceLocator.store
    private val backend = ServiceLocator.backend
    private val queue = ServiceLocator.queue
    private val prefs = ServiceLocator.prefs

    private val _didRestore = MutableStateFlow(false)
    /** 恢复会话与项目之前不画任何一屏，免得登录页闪一下再跳走。 */
    val didRestore: StateFlow<Boolean> = _didRestore.asStateFlow()

    private val _isSignedIn = MutableStateFlow(false)
    val isSignedIn: StateFlow<Boolean> = _isSignedIn.asStateFlow()

    private val _account = MutableStateFlow<AccountUser?>(null)
    /** 本次登录拿到的账号。冷启动只恢复会话不恢复账号（那要多打一次网络），所以可能为 null。 */
    val account: StateFlow<AccountUser?> = _account.asStateFlow()

    private val _selectedProject = MutableStateFlow<RelayProject?>(null)
    val selectedProject: StateFlow<RelayProject?> = _selectedProject.asStateFlow()

    private val _items = MutableStateFlow<List<CaptureItem>>(emptyList())
    /** 全部影像，跨项目。图集切项目、队列页数「别处还有多少」都要看全量。 */
    val items: StateFlow<List<CaptureItem>> = _items.asStateFlow()

    private val _currentItems = MutableStateFlow<List<CaptureItem>>(emptyList())
    val currentItems: StateFlow<List<CaptureItem>> = _currentItems.asStateFlow()

    private val _tally = MutableStateFlow(TransferTally.zero)
    /** 三段计数只数当前项目——进度跟着项目走。 */
    val tally: StateFlow<TransferTally> = _tally.asStateFlow()

    private val _otherPendingCount = MutableStateFlow(0)
    val otherPendingCount: StateFlow<Int> = _otherPendingCount.asStateFlow()

    private val _cloudExpiry = MutableStateFlow<Map<String, String>>(emptyMap())
    /** 未投递件的中转区到期时刻（键 clientMediaId 小写）。队列页做到期提醒用。 */
    val cloudExpiry: StateFlow<Map<String, String>> = _cloudExpiry.asStateFlow()

    private val _desktopOnline = MutableStateFlow(false)
    val desktopOnline: StateFlow<Boolean> = _desktopOnline.asStateFlow()

    val uploadProgress: StateFlow<Map<String, Double>> = queue.progress

    /**
     * 存相册的挂钩，由相机层（Task 8）装上。AppModel 不直接依赖 MediaStore：
     * 相册与上传是两条互不影响的链，任何一条断了另一条都要继续走。
     */
    var albumSaver: ((CaptureItem) -> Unit)? = null

    private var lastProjectsSeenAt: Long? = null
    private var heartbeat: Job? = null

    init {
        // 录音由前台服务落库（用户可能从通知栏按停，界面未必活着），库变了这里跟着重读
        viewModelScope.launch { RecordingState.stored.collect { refresh() } }
    }

    // MARK: - 启动

    fun bootstrap() {
        viewModelScope.launch {
            // 本地有会话就直接进主界面。这里只判断「有没有」，会话是否还有效交给第一次真实
            // 请求的 401——启动多打一次网络请求会让离线开 App 卡住，而离线拍照正是主场景。
            _isSignedIn.value = backend.hasSession()
            _selectedProject.value = prefs.selectedProject
            evidence.sweepOrphans()
            refresh()
            _didRestore.value = true
            if (_selectedProject.value != null) {
                UploadWorker.enqueue(getApplication<Application>())
                _cloudExpiry.value = queue.checkDelivered()
                refresh()
            }
            startHeartbeat()
        }
    }

    /**
     * 上传自愈心跳（dev-board#241 的 Android 对应）：滞留的「传输中」与失败件不能等用户来点，
     * 现场拍完手机就揣兜里了。顺带刷新桌面端在线判定与投递回执。
     */
    private fun startHeartbeat() {
        if (heartbeat?.isActive == true) return
        heartbeat = viewModelScope.launch {
            while (isActive) {
                delay(60_000)
                if (!_isSignedIn.value) continue
                queue.autoKick()
                _cloudExpiry.value = queue.checkDelivered()
                refreshDesktopOnline()
                refresh()
                if (queue.consumeUnauthorized()) signOut()
            }
        }
    }

    /** 项目列表来自桌面端的定时上报：能读到且里面有当前项目那台机器，就认为它此刻开着。 */
    private suspend fun refreshDesktopOnline() {
        val current = _selectedProject.value ?: return
        try {
            val projects = backend.myProjects()
            if (projects.any { it.deviceId == current.deviceId }) lastProjectsSeenAt = System.currentTimeMillis()
        } catch (e: Unauthorized) {
            signOut()
            return
        } catch (_: Exception) {
            // 读不到不等于对面不在——网络抖一下就宣布离线是噪音，交给三分钟窗口自己过期
        }
        _desktopOnline.value = isDesktopOnline(lastProjectsSeenAt, System.currentTimeMillis())
    }

    // MARK: - 账号与项目

    fun didLogin(result: LoginResult) {
        _account.value = result.user
        _isSignedIn.value = true
        startHeartbeat()
        viewModelScope.launch { refresh() }
    }

    fun selectProject(project: RelayProject) {
        prefs.selectedProject = project
        _selectedProject.value = project
        lastProjectsSeenAt = System.currentTimeMillis()
        _desktopOnline.value = true
        viewModelScope.launch {
            refresh()
            UploadWorker.enqueue(getApplication<Application>())
            queue.kick()
        }
    }

    /** 切项目：清掉选择回到选择页。已拍的影像各自记着自己的项目，切项目不改变它们的去向。 */
    fun clearProjectSelection() {
        prefs.selectedProject = null
        _selectedProject.value = null
        _desktopOnline.value = false
        viewModelScope.launch { refresh() }
    }

    /**
     * 退出登录。**不清本地影像**——退出不等于放弃已经拍到的东西，现场是不可复现的，
     * 登出就删是灾难性的默认。
     */
    fun signOut() {
        backend.logout()
        heartbeat?.cancel()
        heartbeat = null
        // 排着的上传任务也撤掉：会话没了再被唤起只会对着每一件打一次 401
        runCatching { WorkManager.getInstance(getApplication()).cancelUniqueWork("upload") }
        _isSignedIn.value = false
        _account.value = null
        prefs.selectedProject = null
        _selectedProject.value = null
        _desktopOnline.value = false
        lastProjectsSeenAt = null
        viewModelScope.launch { refresh() }
    }

    /**
     * 注销账号。删的是**云端**的账号与数据；手机本地的影像不动——理由同 [signOut]。
     * App Store 审核指南 5.1.1(v) 与各安卓应用商店的账号删除要求都指向这个入口。
     */
    suspend fun deleteAccount() {
        backend.deleteAccount()
        signOut()
    }

    /** 云端中转区用量。设置页进页面拉一次；失败让调用方自己决定怎么显示（那里用「—」占位）。 */
    suspend fun mediaUsage(): MediaUsage = backend.mediaUsage()

    // MARK: - 影像

    /**
     * 落一件。顺序是「先进库、再存相册、最后排上传」：库是唯一可信的那一份，
     * 相册与上传都从它派生，任何一条失败都不该回滚已经落好的原件。
     */
    suspend fun store(
        kind: MediaKind,
        tempFile: File,
        capturedAt: Instant,
        loc: Loc?,
        project: RelayProject? = selectedProject.value,
    ): CaptureItem {
        val item = evidence.save(kind, tempFile, capturedAt, loc, project)
        refresh()
        if (prefs.saveToAlbum && kind != MediaKind.audio) albumSaver?.invoke(item)
        UploadWorker.enqueue(getApplication<Application>())
        viewModelScope.launch { queue.kick() }
        return item
    }

    /** 单件重试。队列页每行都有，不用为了一条失败去点「全部重试」。 */
    fun retry(id: String) {
        viewModelScope.launch {
            try {
                evidence.updateState(id, TransferState.waiting, progress = 0.0)
                refresh()
                UploadWorker.enqueue(getApplication<Application>())
                queue.kick()
            } catch (e: Exception) {
                logger.warning("重试 $id 出错: ${e.message}")
            }
        }
    }

    fun retryFailedUploads() {
        viewModelScope.launch {
            try {
                queue.retryFailed()
                refresh()
            } catch (e: Exception) {
                logger.warning("全部重试出错: ${e.message}")
            }
        }
    }

    fun checkDelivered() {
        viewModelScope.launch {
            _cloudExpiry.value = queue.checkDelivered()
            refresh()
        }
    }

    fun delete(ids: Set<String>) {
        viewModelScope.launch {
            try {
                evidence.delete(ids)
                refresh()
            } catch (e: Exception) {
                logger.warning("删除出错: ${e.message}")
            }
        }
    }

    /** 从磁盘重读一遍并派生出各个视图值。读不出来不清空界面——宁可显示上一次的状态。 */
    suspend fun refresh() {
        val all = try { evidence.loadAll() } catch (_: Exception) { return }
        val selected = _selectedProject.value
        _items.value = all
        _currentItems.value = all.filter { it.project?.id == selected?.id }
        _tally.value = TransferTally.of(_currentItems.value)
        _otherPendingCount.value = otherPendingCount(all, selected)
    }
}
